### PR TITLE
[WIP] Release 3.x

### LINKED ISSUE #348: feature: Migrate the codebase to ES2015 (and up)
**Type:** feature

**Description:**

This is a further step to modernizr the codebase. We are using Babel for the UI already and we want to take advance of it for the remaining code. The support old Node4,5 and 6 block us to use the latest features of Javascript. We have been done good progress along the 2017 with the new UI, migrate progressively to ES6 but we have to move on.

The following steps for migration would be:

- [x]  Add type checking (flow) #312
- [x]  Prepare build pipeline for Babel #349
- [x] Unit testing must support ES6 with the current set up, minor changes  #349
- [x] Create type definitions for flow (https://github.com/verdaccio/flow-types)
- [x] Update Docker build image for transpiling
- [x] Code coverage has to support ES6
- [x] Local storage must be extracted to separate repository using ES6 #291 

After these steps, the migration can continue progressively until is done.

### PR DESCRIPTION
**Pull request by juanpicado**
**Type:** feature

Future major release.

 - [x] [MIT](https://opensource.org/licenses/MIT) will be the default license (https://github.com/verdaccio/verdaccio/pull/446)

Work in progress.

## New Features
- [ ] Verdaccio is cluster ready #103 @juanpicado 
- [ ] Improve plugin system #169 @juanpicado
- [ ] Allow users change password via UI #401 (PR-Welcome)
- [x] New side bar (package info) to detail page (UI) (#416)

*I've moved some features for future minor releases thus this will speed up this big release*

## Development 
- [x] Migrate from Mocha to Jest #406
- [x] Jest unit test for UI #415 
- [x] Docker should lint and run unit tests on build the image 
- [x] Migrate the codebase to ES2015 (and up) #348 
- [ ] Extract built-in htpasswd as external plugin dependency #192  @ayusharma [WIP]
- [x] We drop Node4 support, latest Node 6 is the minimum #419 
- More enhancements to come ...

## Website
- [ ] We are gonna roll out a new website soon @rodriguesbreno powered by (https://docusaurus.io/) #439 #386
- [x] Twitter looks and feel
- [x] Avatar Docker Hub, Github, Twitter and Gravatar
- [x] Cloudron Logo (https://cloudron.io/store/org.eggertsson.verdaccio.html)

⚠️  All future features should branch from `3.x`
- If you want to contribute just leave a comment for coordination

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -23,7 +23,6 @@ RUN npm config set registry http://registry.npmjs.org/ && \
     npm install -g -s flow-bin@0.60.0 && \
     yarn install --production=false && \
     yarn run lint && \
-    yarn run test && \
```
Comment (ayusharma) [inline Dockerfile]: I'm still figuring out why tests are failing on docker :( I hope to find a solution in 2-3 days.

Comment (juanpicado) [reply Dockerfile]: Sure, no rush, I disabled them now temporary. I hope you find a reason ;( I couldn't.

Comment (juanpicado) [reply Dockerfile]: https://github.com/verdaccio/verdaccio/commit/c2780d6746cd8115c4e1cc48f3a09c688b8d6fb9 
From the slack channel for the audience.
> "react does not throw errors on Production mode in Docker container we switch the environment to Production, to fix it, I just reset the env to something different and now they run fine"

**Comment by StarpTech** [comment]
What about following changes:

- Support only Node >= 8 to use async / await ?
- Don't split core packages in multiple packages (`@verdaccio/*`) when they aren't used somewhere else.
- Use prettier as code formatting tool.

**Comment by juanpicado** [comment]
What about following changes:

> Support only Node >= 8 to use async / await ?

We transpile with Babel and Docker image use Node9

> Don't split core packages in multiple packages (@verdaccio/*) when they aren't used somewhere else.

There is a reason for this, new plugin system.

> Use prettier as code formatting tool. 

It was dropped https://github.com/verdaccio/verdaccio/pull/380

**Comment by StarpTech** [comment]
> It was dropped #380

I can't imagine that all these [users](https://prettier.io/en/users/) do it wrong. In my opinion, we should give it a try.

**Comment by juanpicado** [comment]
I haven't tried to be honest, perhaps when the queue is a bit empty. I think @ayusharma has better idea in this field than me.

**Comment by juanpicado** [comment]
We will finish the remaining task in separated PR

**Pull request merged**